pgx/src/heap_tuple.rs

```diff
@@ -0,0 +1,340 @@
+        data.t_data = htup_header;
+
+        Self {
+            tuple: data,
+            tupdesc: PgTupleDesc::from_pg(tupdesc),
+        }
+    }
+
+    /// Given the name for an attribute in this [PgHeapTuple], change its value.
+    ///
+    /// Attribute names are case sensitive.
+    ///
+    /// ## Errors
+    ///
+    /// - return [TryFromDatumError::NoSuchAttributeName] if the attribute does not exist
+    /// - return [TryFromDatumError::IncompatibleTypes] if the Rust type of the `value` is not
+    /// compatible with the attribute's Postgres type
+    pub fn set_by_name<T: IntoDatum>(&mut self, attname: &str, value: T) -> FromDatumResult<()> {
+        match self.get_attribute_by_name(attname) {
+            None => Err(TryFromDatumError::NoSuchAttributeName(attname.to_string())),
+            Some((attnum, _)) => self.set_by_index(attnum, value),
+        }
+    }
+
+    /// Given the index for an attribute in this [PgHeapTuple], change its value.
+    ///
+    /// Attribute numbers start at 1, not 0.
+    ///
+    /// ## Errors
+    /// - return [TryFromDatumError::NoSuchAttributeNumber] if the attribute does not exist
+    /// - return [TryFromDatumError::IncompatibleTypes] if the Rust type of the `value` is not
+    /// compatible with the attribute's Postgres type
+    pub fn set_by_index<T: IntoDatum>(
+        &mut self,
+        attno: NonZeroUsize,
+        value: T,
+    ) -> FromDatumResult<()> {
+        unsafe {
+            match self.get_attribute_by_index(attno) {
+                None => return Err(TryFromDatumError::NoSuchAttributeNumber(attno)),
+                Some(att) => {
+                    if !T::is_compatible_with(att.atttypid) {
+                        return Err(TryFromDatumError::IncompatibleTypes);
+                    }
+                }
+            }
+
+            let mut datums = (0..self.tupdesc.len()).collect::<Vec<_>>();
+            let mut nulls = (0..self.tupdesc.len()).map(|_| false).collect::<Vec<_>>();
+            let mut do_replace = (0..self.tupdesc.len()).map(|_| false).collect::<Vec<_>>();
+
+            let datum = value.into_datum();
+            let attno = attno.get() - 1;
+
+            nulls[attno] = datum.is_none();
+            datums[attno] = datum.unwrap_or(0);
+            do_replace[attno] = true;
+
+            let new_tuple = PgBox::<pg_sys::HeapTupleData, AllocatedByRust>::from_rust(
+                pg_sys::heap_modify_tuple(
+                    self.tuple.as_ptr(),
+                    self.tupdesc.as_ptr(),
+                    datums.as_mut_ptr(),
+                    nulls.as_mut_ptr(),
+                    do_replace.as_mut_ptr(),
+                ),
+            );
+            let old_tuple = std::mem::replace(&mut self.tuple, new_tuple);
+            drop(old_tuple);
+            Ok(Some(()))
+        }
+    }
+}
+
+impl<'a, AllocatedBy: WhoAllocated<pg_sys::HeapTupleData>> PgHeapTuple<'a, AllocatedBy> {
```

Comment:
Noting for @workingjubilee this is a similar pattern to how `PgBox` works.

Reply:
That's really poorly named.  Instead of `WhoAllocated` it should be `WhoNeedsToFreeThisThing`.

The reason is, despite if it's `AllocatedByRust` or `AllocatedByPostgres`, it's a postgres `palloc()`'d value.

We can address this at a later time.  Not relevant to this PR.

Reply:
WhoDrops

Reply:
Ya know, @workingjubilee, why don't you put up a PR for that (just off of current `develop` branch) and otherwise give `PgBox` an audit that only someone with your skills have.

It's looking like our next release is going to have some pretty big API breakage anyways, so now would be a good time to rejigger naming things...